x/qgb/abci.go

```diff
@@ -113,3 +122,108 @@ func handleValsetRequest(ctx sdk.Context, k keeper.Keeper) {
 		}
 	}
 }
+
+func PruneIfNeeded(ctx sdk.Context, k keeper.Keeper) {
+	if !k.CheckLatestAttestationNonce(ctx) ||
+		k.CheckLatestAttestationNonce(ctx) && k.GetLatestAttestationNonce(ctx) <= AttestationPruningThreshold {
+		// If the attestations nonce hasn't been initialized yet, or the nonce is not greater than the minimum
+		// number of attestations, we don't need to prune anything
+		return
+	}
+
+	lastAvailableNonce := k.GetLastPrunedAttestationNonce(ctx) + 1
+	lastAttestationInStore, found, err := k.GetAttestationByNonce(ctx, lastAvailableNonce)
+	if err != nil {
+		panic(err)
+	}
+	if !found {
+		panic(fmt.Sprintf("couldn't find attestation %s for pruning", lastAttestationInStore))
+	}
+	var lastNonceHeight uint64
+	switch lastAttestationInStore.Type() {
+	case types.DataCommitmentRequestType:
+		dc, ok := lastAttestationInStore.(*types.DataCommitment)
+		if !ok {
+			panic("couldn't cast data commitment type")
+		}
+		lastNonceHeight = dc.EndBlock
+	case types.ValsetRequestType:
+		vs, ok := lastAttestationInStore.(*types.Valset)
+		if !ok {
+			panic("couldn't cast valset type")
+		}
+		lastNonceHeight = vs.Height
+	}
+	lastUnbondingHeight := k.GetLastUnBondingBlockHeight(ctx)
```

Comment:
there are multiple unbonding heights in the history, this one refers to the last one

Reply:
I think we inherited this portion from the gravity module, but are we using the validator's unbonding height as a proxy for the real unbonding height (the height from 3 weeks previous)? If so we might have to change that, as the important thing is that 3 weeks value, we don't want to prune after that

Reply:
Actually, that value is set in here: https://github.com/celestiaorg/celestia-app/blob/5443e5fca8fb02652dafbca8a64a62c8263bfe96/x/qgb/keeper/hooks.go#L32

Which changes on every validator that starts unbonding. So for example, if validator `A` starts unbonding on height 100, it will take that value. Then, if validator `B` starts unbonding at height 110, it will update its value to 110.

Not sure about the above scenario, if any validator can unbound anytime they want (I assume yes) or not